sys/src/9/port/devether.c

```diff
@@ -176,36 +185,40 @@ etheriq(Ether* ether, Block* bp, int fromwire)
 	fromme = memcmp(pkt->s, ether->ea, sizeof(pkt->s)) == 0;
 
 	/*
-	 * Multiplex the packet to all the connections which want it.
+	 * Multiplex the packet to all the connections that want it.
 	 * If the packet is not to be used subsequently (fromwire != 0),
 	 * attempt to simply pass it into one of the connections, thereby
 	 * saving a copy of the data (usual case hopefully).
 	 */
 	for(fp = ether->Netif.f; fp < ep; fp++){
-		if(f = *fp)
+		f = *fp;
+		if(f != nil)
 		if(f->type == type || f->type < 0)
-		if(tome || multi || f->prom || f->bridge & 2){
+		if(tome || multi || f->prom || (f->bridge & 0x02)){
 			/* Don't want to hear bridged packets */
 			if(f->bridge && !fromwire && !fromme)
 				continue;
-			if(!f->headersonly){
-				if(fromwire && fx == 0)
-					fx = f;
-				else if(xbp = iallocb(len)){
-					memmove(xbp->wp, pkt, len);
-					xbp->wp += len;
-					if(qpass(f->iq, xbp) < 0)
-						ether->Netif.soverflows++;
-				}
-				else
-					ether->Netif.soverflows++;
-			}
-			else
+			if(f->headersonly){
 				etherrtrace(f, pkt, len);
+				continue;
+			}
+			if(fromwire && fx == 0){
```

Comment:
I expect so. Barret and I talked about this once vis-a-vis Akaros, and the idea was that they may have been trying to force more efficient code generation. I think we can get around that by doing an FDO-style build.